Task2/Project/Shapes/Models/Canvas.cs

```diff
@@ -2,106 +2,230 @@
     {
+        // FIELDS
+        List<ShapeBase> shapes;
+
+        // CONSTRUCTORS
+        /// <summary>
+        /// Basic constructor without parameters
+        /// </summary>
+        public Canvas()
+        {
+            shapes = new List<ShapeBase>();
+        }
+
         // PROPERTIES
+        /// <summary>
+        /// Property that enable to get count of shapes
+        /// </summary>
+        /// /// <returns>Count of shapes</returns>
         public int Count
         {
             get
             {
-                throw new NotImplementedException();
+                return shapes.Count;
             }
         }
+        /// <summary>
+        /// Property that shows if canvas is only for read
+        /// </summary>
+        /// /// <returns>Is canvas only for read</returns>
         public bool IsReadOnly
         {
             get
             {
-                throw new NotImplementedException();
+                return false;
             }
         }
-        
+        /// <summary>
+        /// Property that enable to get shapes collection
+        /// </summary>
+        /// /// <returns>Shapes collecton</returns>
         public IEnumerable<ShapeBase> Shapes
         {
             get
             {
-                throw new NotImplementedException();
+                return shapes;
             }
         }
+
         // INDEXERS
+        /// <summary>
+        /// Indexer that enable to interract with collection elements
+        /// </summary>
+        /// <param name="index">Shape index in collection</param>
+        /// <returns>Shape with preset index</returns>
+        /// <exception cref="System.IndexOutOfRangeException">Index out of range</exception>
         public ShapeBase this[int index]
         {
             get
             {
-                throw new NotImplementedException();
+                return shapes[index];
             }
-
             set
             {
-                throw new NotImplementedException();
+                shapes[index].PropertyChanged -= Canvas_PropertyChanged;
+                shapes[index] = value;
+                shapes[index].PropertyChanged += Canvas_PropertyChanged;
+                OnCollectionChanged(new NotifyCollectionChangedEventArgs(NotifyCollectionChangedAction.Replace));
             }
         }
 
         // EVENTS
+        /// <summary>
+        /// Event that message about canvas property change
+        /// </summary>
         public event NotifyCollectionChangedEventHandler CollectionChanged;
 
         // METHODS
+        /// <summary>
+        /// Method that allow to add new shape to collection
+        /// </summary>
+        /// <param name="shape">Shape that should be added</param>
         public void Add(ShapeBase shape)
         {
-            throw new NotImplementedException();
+            shapes.Add(shape);
+            shapes[Count - 1].PropertyChanged += Canvas_PropertyChanged;

```

Comment:
**shapes[Count - 1]** hard to read. just use **shape** that you have in argument

Reply:
but its different objects...

Reply:
no